### PR TITLE
Handle `depth` in all readDirectory implementations

### PR DESCRIPTION
**Pull request by andy-ms**
`discoverTypings` needs to pass a `depth` parameter to `readDirectory` to ensure that `node_modules` is not traversed deeply. (This may have caused the slowdown in #16203.) Unfortunately, TypeScript allowed us to write an optional `depth` parameter at the declaration, even if it doesn't appear in almost any of the implementations. This PR adds `depth` to all of our implementations.

It's possible that third party implementations still don't handle a `depth` parameter; don't think there are any third-party hosts being used for `discoverTypings` though.

It's missing support for `ChakraHost` though. @paulvanbrenk How would we add support for that?

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -2020,7 +2021,7 @@ namespace ts {
         };
     }
 
-    export function matchFiles(path: string, extensions: string[], excludes: string[], includes: string[], useCaseSensitiveFileNames: boolean, currentDirectory: string, getFileSystemEntries: (path: string) => FileSystemEntries): string[] {
+    export function matchFiles(path: string, extensions: string[], excludes: string[], includes: string[], useCaseSensitiveFileNames: boolean, currentDirectory: string, depth: number | undefined, getFileSystemEntries: (path: string) => FileSystemEntries): string[] {
```
Comment (jramsay) [inline src/compiler/core.ts]: Is this a potential breaking change if anyone else is using matchFiles? Should depth be added as the final param?

Comment (andy-ms) [inline src/compiler/core.ts]: This is `/* @internal */`. (See comment at top of the enclosing namespace)

File: src/compiler/core.ts [hunk 2]
```diff
@@ -2,8 +2,9 @@
 /// <reference path="performance.ts" />
 
 namespace ts {
+    export const versionMajorMinor = "2.5";
     /** The version of the TypeScript compiler release */
-    export const version = "2.5.0";
+    export const version = `${versionMajorMinor}.0`;
```
Comment (mihailik) [inline src/compiler/core.ts]: That escalated quickly!

Whence has this change come from?

File: src/services/shims.ts [hunk 3]
```diff
@@ -467,7 +467,7 @@ namespace ts {
         }
     }
 
-    export class CoreServicesShimHostAdapter implements ParseConfigHost, ModuleResolutionHost {
+    export class CoreServicesShimHostAdapter implements ParseConfigHost, ModuleResolutionHost, JsTyping.TypingResolutionHost {
```
Comment (mihailik) [inline src/services/shims.ts]: Another seemingly unrelated change in this PR?

Is this a driveby bugfix, or actually related to the `depth` issue?

Comment (andy-ms) [inline src/services/shims.ts]: It always implemented that interface. TypeScript's structural type system doesn't require us to declare these things, but it doesn't hurt.

File: src/compiler/core.ts (hunk 2 above)
Comment (andy-ms) [inline src/compiler/core.ts]: This shouldn't actually change the value. I needed to access just the `major.minor` version in a test, so I made a variable just for that, and wrote `version` in terms of it so they wouldn't get out-of-sync.

Comment (mihailik) [inline src/compiler/core.ts]: Ah, see that now, thanks a lot!

I wonder whether moving complexity out of testing code into the compiler is worthy. Especially, now it's surfaced in the API in a confusing way (can we still assume normal version is dot-joined numbers?). And of course readability and text-searchability suffers.

Comment (andy-ms) [inline src/compiler/core.ts]: versionMajorMinor was required in several places (not just tests), but previously we were parsing it out of `version`. Figured it would be better to construct `version` from `versionMajorMinor` than the other way around.

File: src/compiler/core.ts (hunk 1 above)
Comment (rwjblue) [inline src/compiler/core.ts]: What was the reason for intentionally breaking the signature here (the only answer above to the question is that it is supposedly private)? Regardless of the `/* @internal */` comment this method is exported (without an underscore or other indication of it being private) on the main export (e.g. `import { matchFiles } from 'typescript'`). 

This is definitely a breaking change and makes life difficult for tooling that must support both 2.4 and 2.5.  See https://github.com/tildeio/broccoli-typescript-compiler/issues/49.

Comment (andy-ms) [inline src/compiler/core.ts]: Ref: #17891
